test/e2e/storage/csimock/csi_node_volume_health.go

```diff
@@ -0,0 +1,199 @@
+			test := t
+			ginkgo.It(test.name, func(ctx context.Context) {
+				m.init(ctx, testParameters{
+					registerDriver:            true,
+					enableNodeVolumeCondition: test.nodeVolumeConditionRequired,
+					hooks:                     createGetVolumeStatsHook(test.nodeAbnormalVolumeCondition),
+				})
+				ginkgo.DeferCleanup(m.cleanup)
+				_, claim, pod := m.createPod(ctx, pvcReference)
+				if pod == nil {
+					return
+				}
+				// Wait for PVC to get bound to make sure the CSI driver is fully started.
+				err := e2epv.WaitForPersistentVolumeClaimPhase(ctx, v1.ClaimBound, f.ClientSet, f.Namespace.Name, claim.Name, time.Second, framework.ClaimProvisionTimeout)
+				framework.ExpectNoError(err, "while waiting for PVC to get provisioned")
+
+				ginkgo.By("Waiting for pod to be running")
+				err = e2epod.WaitForPodNameRunningInNamespace(ctx, m.cs, pod.Name, pod.Namespace)
+				framework.ExpectNoError(err, "wait for running pod")
+				ginkgo.By("Waiting for all remaining expected CSI calls")
+				err = wait.PollUntilContextTimeout(ctx, time.Second, csiNodeVolumeStatWaitPeriod, true, func(c context.Context) (done bool, err error) {
+					var index int
+					_, index, err = compareCSICalls(ctx, trackedCalls, test.expectedCalls, m.driver.GetCalls)
+					if err != nil {
+						return true, err
+					}
+					if index == 0 {
+						// No CSI call received yet
+						return false, nil
+					}
+					if len(test.expectedCalls) == index {
+						// all calls received
+						return true, nil
+					}
+					return false, nil
+				})
+				// try to use ```csi.NewMetricsCsi(pv.handler).GetMetrics()``` to get metrics from csimock driver but failed.
+				// the mocked csidriver register doesn't regist itself to normal csidriver.
+				if test.nodeVolumeConditionRequired {
+					pod, err := f.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+					framework.ExpectNoError(err, "get running pod")
+					grabber, err := e2emetrics.NewMetricsGrabber(ctx, f.ClientSet, nil, f.ClientConfig(), true, false, false, false, false, false)
+					framework.ExpectNoError(err, "creating the metrics grabber")
+					waitErr := wait.PollUntilContextTimeout(ctx, 30*time.Second, csiNodeVolumeStatWaitPeriod, true, func(ctx context.Context) (bool, error) {
+						framework.Logf("Grabbing Kubelet metrics")
+						// Grab kubelet metrics from the node the pod was scheduled on
+						var err error
+						kubeMetrics, err := grabber.GrabFromKubelet(ctx, pod.Spec.NodeName)
+						if err != nil {
+							framework.Logf("Error fetching kubelet metrics err: %v", err)
+							return false, err
+						}
+						if !findVolumeConditionMetrics(f.Namespace.Name, claim.Name, kubeMetrics, test.nodeAbnormalVolumeCondition) {
+							return false, nil
+						}
+						return true, nil
+					})
+					framework.ExpectNoError(waitErr, "call metrics should not have any error")
+				}
+				framework.ExpectNoError(err, "while waiting for all CSI calls")
```

Comment:
Why is this not after the call which sets `err`?

Reply:
fixed

Reply:
@pohly please take a look again, all comments are fixed